src/Controller/SantaController.php

```diff
@@ -90,25 +92,36 @@ public function participants(Request $request, string $application): Response
         $selectedUsers = $session->get('selected-users', []);
         $errors = [];
 
+        $form = $this->createForm(ParticipantType::class, null, [
+            'available-users' => $availableUsers,
+        ]);
+
         if ($request->isMethod('POST')) {
-            $selectedUsers = $request->request->all('users');
+            $form->handleRequest($request);
 
-            if (\count($selectedUsers) > 1) {
-                $session->set('selected-users', $selectedUsers);
+            // looks counterintuitive, but it ensures that the symfony form doesn't get sent again when going back on the page from the next step
+            $selectedUsers = [];
 
-                return $this->redirectToRoute('message', ['application' => $application->getCode()]);
-            }
+            if ($form->isSubmitted() && $form->isValid()) {
+                foreach ($form->getData()['users'] as $user) {
+                    $selectedUsers[] = $user->getIdentifier();
+                }
 
+                if (\count($selectedUsers) > 1) {
+                    $session->set('selected-users', $selectedUsers);
+
+                    return $this->redirectToRoute('message', ['application' => $application->getCode()]);
+                }
+            }
             $errors[] = 'At least 2 users should be selected.';
         }
 
-        $content = $this->twig->render('santa/application/participants_' . $application->getCode() . '.html.twig', [
-            'application' => $application->getCode(),
+        $content = $this->twig->render('santa/application/participants_' . $application->getCode() . '.html.twig', ['application' => $application->getCode(),
             'users' => $availableUsers,
             'groups' => $application->getGroups(),
             'selectedUsers' => $selectedUsers,
-            'errors' => $errors,
-        ]);
+            'form' => $form->createView(),
+            'errors' => $errors, ]);
```

Comment:
this `errors` variable can be replaced by retrieving errors directly from the form